classes/local/filter_base.php

```diff
@@ -0,0 +1,90 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * An abstract class to generate rubrics based on config values.
+ * Subclasses will need to define get_rubrics, and may wish to subclass
+ * config_validator and reference that subclass from $configvalidator.
+ *
+ * @package    block_filtered_course_list
+ * @copyright  2016 CLAMP
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+namespace block_filtered_course_list\local;
+
+defined('MOODLE_INTERNAL') || die();
+
+abstract class filter_base {
+    /** @var array Elements of the rubric config line */
+    public $config;
+    /** @var array A list of courses the current user is enrolled in */
+    public $courselist;
+    /** @var array Config settings for the block */
+    public $blockconfig;
+    /** @var array A list of the rubric objects generated by the line */
+    public $rubrics = array();
+
+    /**
+     * Constructor
+     *
+     * @param array $config Array of config elements parsed from line
+     * @param array $courselist An array of courses the user is enrolled in
+     * @param array $blockconfig Details of the block configuration
+     */
+    public function __construct($config, $courselist, $blockconfig) {
+        $this->courselist = $courselist;
+        $this->blockconfig = $blockconfig;
+        $this->config = $this->validate_config($config);
+    }
+
+    /**
+     * Run config validation on given config data, using the config validator
+     * for this subclass, if it exists, or the nearest defined config validator
+     * going up the class tree.
+     * This method is static so that we can run it up through parent classes
+     * without instantiating.
+     *
+     * @return string \block_filtered_course_list\local\filter_{filtername}\config_validator
+     */
+    public static function get_config_validator() {
+        $classname = get_called_class();
+        $namespace = preg_replace('/\\w+$/', '', $classname);
+        $validator = "$namespace" . "config_validator";
+        if (class_exists($validator)) {
+            return $validator;
+        } elseif ($pclass = get_parent_class(get_called_class())) {
+            return $pclass::get_config_validator();
+        }
+    }
```

Comment:
It feels to me like you are recreating the inheritance behavior that one gets automatically from extending a class, and I'm not sure why we can't just rely on inheritance. I also feel like we are losing the benefit of defining an abstract method. The point of the abstract method is to compel a child class to define this function for itself. By defining an explicit function here, aren't we letting the child classes off the hook in a way that we probably don't want to do?

Reply:
I added this to replicate the current functionality of the regex filter. The `regex_configline` class inherits from the `shortname_configclass`, and uses its `validate_line()` method.

In the same way, in the new structure, it should use the same config validator, since it doesn't define any config of its own. 

That said, I agree with you, and frankly I can't remember why I ended up doing it that way in the first place. I think it would be better to have each class define a property "config_validator" or "config_validator_name" -- then each subclass can point at whichever one it wants. I suppose that makes it slightly less automatic in terms of creating new filters, but still probably a better pattern.

Reply:
Ultimately I think that an even better pattern would be a single config class, with which each filter plugin can register config values -- say, by passing name, default value, and type or regex match. For example, it might look like `$config->add_field('label', '/\w+/', 'Default Label')`

That would have a few benefits:
1. Separating config validation logic into a single place
2. Allowing each filter to declare custom config fields with a suitably small amount of code
3. Potentially allowing us to drop the complex classes directory structure in favor of single files in the base `classes` directory